### PR TITLE
Write amqp1 plugin

### PR DESCRIPTION
**Pull request by ajssmith**
This write plugin allows data to be sent to an AMQP 1.0 intermediary such as the Apache Qpid Dispatch Router (e.g. direct messaging) or Apache Artemis Broker (e.g. queue backed messaging).

It is similar to the existing amqp plugin which is for AMQP version 0.9.1 (e.g. rabbitmq server).

### CODE REVIEW

**Comment by leifmadsen** [comment]
Just realized that you also need to adjust the RPM spec file so that the module can be added to the RPM.

https://github.com/collectd/collectd/blob/master/contrib/redhat/collectd.spec

File: Makefile.am [hunk 1]
```diff
@@ -542,6 +542,18 @@ amqp_la_LIBADD = \
 	libformat_json.la
 endif
 
+if BUILD_PLUGIN_AMQP1
+pkglib_LTLIBRARIES += amqp1.la
+amqp1_la_SOURCES = src/amqp1.c
```
Comment (leifmadsen) [inline Makefile.am]: You also need to add `src/utils_deq.h` or else it won't be included in the tarball which is used for building RPMs.

```
amqp1_la_SOURCES = src/amqp1.c \
        src/utils_deq.h
```

**Comment by bendoerr** [comment]
+1 for this! I've got [prior art](https://github.com/bendoerr/collectd-write-qpid-json) in the form of a python plugin, but this on the surface looks better.

File: configure.ac [hunk 2]
```diff
@@ -4686,6 +4686,47 @@ if test "$with_libpython" != "xno"; then
...
+
+  AC_CHECK_HEADERS([proton/proactor.h],
+    [with_libqpid_proton="yes"],
+    [with_libqpid_proton="no (proton/proactor.h not found)"]
+  )
+
+  CPPFLAGS="$SAVE_CPPFLAGS"
+fi
+
+# add library checks for function calls
```
Comment (rubenk) [inline configure.ac]: would be nice to indeed check if you can link against the library.

File: src/amqp1.c [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define BUFSIZE 8192
+#define AMQP1_FORMAT_JSON 0
+#define AMQP1_FORMAT_COMMAND 1
+#define AMQP1_FORMAT_GRAPHITE 2
+
+typedef struct amqp1_config_transport_t {
```
Comment (rubenk) [inline src/amqp1.c]: we use _s for the struct tagname, _t for the typedef as a convention

File: src/amqp1.c [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+ * Globals
+ */
+pn_connection_t *conn = NULL;
+pn_link_t *sender = NULL;
+pn_proactor_t *proactor = NULL;
+pthread_mutex_t send_lock;
+cd_message_list_t out_messages;
+uint64_t cd_tag = 1;
+uint64_t acknowledged = 0;
+amqp1_config_transport_t *transport = NULL;
```
Comment (rubenk) [inline src/amqp1.c]: can these be made static?

File: src/amqp1.c [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+static bool stopping = false;
+static int event_thread_running = 0;
+static pthread_t event_thread_id;
+
+/*
+ * Functions
+ */
+static void cd_message_free(cd_message_t *cdm) {
+  if (cdm->mbuf.start) {
+    free((void *)cdm->mbuf.start);
```
Comment (rubenk) [inline src/amqp1.c]: unnecessary cast

File: src/amqp1.c [hunk 6]
```diff
@@ -0,0 +1,730 @@
...
+
+static bool stopping = false;
+static int event_thread_running = 0;
+static pthread_t event_thread_id;
+
+/*
+ * Functions
+ */
+static void cd_message_free(cd_message_t *cdm) {
+  if (cdm->mbuf.start) {
```
Comment (rubenk) [inline src/amqp1.c]: free accepts NULL so no need for the check

File: src/amqp1.c [hunk 7]
```diff
@@ -0,0 +1,730 @@
...
+  char *user;
+  char *password;
+  char *address;
+  int retry_delay;
+} amqp1_config_transport_t;
+
+typedef struct amqp1_config_instance_t {
+  DEQ_LINKS(struct amqp1_config_instance_t);
+  char *name;
+  _Bool notify;
```
Comment (rubenk) [inline src/amqp1.c]: It's fine to use bool from <stdbool.h>

File: src/amqp1.c [hunk 8]
```diff
@@ -0,0 +1,730 @@
...
+  if ((n == NULL) || (user_data == NULL))
+    return EINVAL;
+
+  instance = user_data->data;
+
+  if (instance->notify != true) {
+    ERROR("amqp1 plugin: write notification failed");
+  }
+
+  cdm = NEW(cd_message_t);
```
Comment (rubenk) [inline src/amqp1.c]: I'd prefer it if you just use malloc / calloc.

File: src/amqp1.c [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+    if (status != 0)
+      break;
+  }
+
+  if (status != 0) {
+    amqp1_config_instance_free(instance);
+    return status;
+  } else {
+    char tpname[1024];
+    int status;
```
Comment (rubenk) [inline src/amqp1.c]: this shadows status from the outer scope

File: src/amqp1.c [hunk 10]
```diff
@@ -0,0 +1,730 @@
...
+              child->key);
+    if (status != 0)
+      break;
+  }
+
+  if (status != 0) {
+    amqp1_config_instance_free(instance);
+    return status;
+  } else {
+    char tpname[1024];
```
Comment (rubenk) [inline src/amqp1.c]: why 1024? we normally use DATA_MAX_NAME_LEN I believe.

File: src/amqp1.c [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+  int status = 0;
+
+  transport = calloc(1, sizeof(*transport));
+  if (transport == NULL) {
+    ERROR("amqp1 plugin: calloc failed.");
+    return ENOMEM;
+  }
+
+  /* Initialize transport configuration {{{ */
+  transport->name = NULL;
```
Comment (rubenk) [inline src/amqp1.c]: No need to set this, calloc has already zero-initialized it.

File: src/amqp1.c [hunk 12]
```diff
@@ -0,0 +1,730 @@
...
+static int amqp1_config(oconfig_item_t *ci) /* {{{ */
+{
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+
+    if (strcasecmp("Transport", child->key) == 0)
+      amqp1_config_transport(child);
+    else
+      WARNING("amqp1 plugin: Ignoring unknown config iption \%s\".",
```
Comment (rubenk) [inline src/amqp1.c]: s/iption/option/

File: src/amqp1.c (hunk 7 above)
Comment (MariuszX) [reply src/amqp1.c]: in whole code please migrate from _Bool to bool

File: src/amqp1.c [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+      status = cf_util_get_flag(child, &instance->graphite_flags,
+                                GRAPHITE_PRESERVE_SEPARATOR);
+    else if (strcasecmp("GraphitePrefix", child->key) == 0)
+      status = cf_util_get_string(child, &instance->prefix);
+    else if (strcasecmp("GraphitePostfix", child->key) == 0)
+      status = cf_util_get_string(child, &instance->postfix);
+    else if (strcasecmp("GraphiteEscapeChar", child->key) == 0) {
+      char *tmp_buff = NULL;
+      status = cf_util_get_string(child, &tmp_buff);
+      if (strlen(tmp_buff) > 1)
```
Comment (MariuszX) [inline src/amqp1.c]: Check status before. tmp_buff could still be NULL here on incorrect data in config file.

File: src/amqp1.c [hunk 14]
```diff
@@ -0,0 +1,730 @@
...
+  amqp1_config_instance_t *instance;
+
+  instance = calloc(1, sizeof(*instance));
+  if (instance == NULL) {
+    ERROR("amqp1 plugin: calloc failed.");
+    return ENOMEM;
+  }
+
+  /* Initialize instance configuration {{{ */
+  instance->name = NULL;
```
Comment (MariuszX) [inline src/amqp1.c]: No need to set this, calloc has already zero-initialized it.

File: src/amqp1.c [hunk 15]
```diff
@@ -0,0 +1,730 @@
...
+  char *name;
+  char *host;
+  char *port;
+  char *user;
+  char *password;
+  char *address;
+  int retry_delay;
+} amqp1_config_transport_t;
+
+typedef struct amqp1_config_instance_t {
```
Comment (MariuszX) [inline src/amqp1.c]: use _s for the struct tagname (same as for amqp1_config_transport_t  above)

**Comment by ajssmith** [comment]
Thanks for taking the time to review and provide comments. I have made the updates to the plugin and look forward to any additional feedback and comments.

**Comment by paramite** [comment]
+1

**Comment by rubenk** [comment]
Thank you for your patience @ajssmith. Merged into master.

**Comment by paramite** [comment]
Thanks a lot @rubenk !

**Comment by ajssmith** [comment]
@rubenk Thanks!

**Pull request merged**